keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/optimizer/gradient/GradientOptimizer.java

```diff
@@ -40,6 +41,15 @@
     }
 
     public static GradientOptimizer of(BayesianNetwork bayesNet) {
+        long numIntegerVertices = bayesNet.getDiscreteLatentVertices().stream()
+            .filter(v -> v instanceof IntegerVertex)
```

Comment:
is Integer vertex the only type of discretes we have? What about something like a categorical vertex?

Reply:
also could extract this condition to a function called "isDiscreteVertex", which would make this more explicit, and also allow us to change the criteria for how we identify discrete vertices

Reply:
If bayesNet.getDiscreteLatentVertices() returns anything other than an empty list then we should throw. Not need to filter for just integers.